### PR TITLE
Add support for local and global chats in Discord

### LINKED ISSUE #4989: EssentialsDiscord: local messages treated as global and relayed when all players are in local message range
### Type of bug

Exploit

### `/ess dump all` output

https://essentialsx.net/dump.html?id=5b9d8973f60f490b9ccedf0513ffbc1b

### Error log (if applicable)

_No response_

### Bug description

Local messages treated as global and relayed when all players are in local message range.
Yes, I am using Mohist, but the problem is in Essentials code.

https://github.com/EssentialsX/Essentials/blob/d86b471402e20f39a774fba2ae5ba6dc3cc26712/EssentialsDiscord/src/main/java/net/essentialsx/discord/listeners/BukkitListener.java#L88

### Steps to reproduce

1. Setup discord channel
2. Configurate `EssentialsDiscord\config.yml`, set `chat.show-all-chat` to `false`
3. Have two people in local message range
4. One player should say something (as local message)

### Expected behaviour

Message should not be relayed to Discord chat

### Actual behaviour

Message is relayed to Discord chat

### PR DESCRIPTION
**Pull request by brawaru**
**This pull request is a draft since it requires #4683 to be merged first.**

***

### Information

×

### Details

**Proposed feature:**  

This change makes use of added separate events for local and global chats in , making it possible to change the formatting of messages sent from different chat types, as well as disable them altogether (for example, if you don't want to see local chat messages in Discord).

By default Essentials Discord will continue to use Bukkit chat events without any kind of separation, you must explicitly opt in to have new behaviour by changing use-essential-events to true.

**Environments tested:**    

OS: Windows 10 20H2.

Java version:
```
openjdk version "17" 2021-09-14 LTS
OpenJDK Runtime Environment Zulu17.28+13-CA (build 17+35-LTS)
OpenJDK 64-Bit Server VM Zulu17.28+13-CA (build 17+35-LTS, mixed mode, sharing)
```

- [x] Most recent Paper version (1.18.1, git-Paper-71)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**  

![Screenshot of the Minecraft chat mirroring to Discord. Brawaru joins the server and says ‘hey’, it gets repeated to Discord as ‘Brawaru: hey’. Brawaru then says ‘now I will enable essentials events’. After enabling them Brawaru says ‘nothing has changed, since local chat is not enabled’, continued ‘I will now enable local chat’. Those messages were repeated to Discord as before. Brawaru then says ‘hello from local chat’, an ‘L’ appears at the beginning of his message, the message gets repeated in Discord as ‘[Local] Brawaru: hello from local chat’. Brawaru then uses question syntax and asks, ‘I am rather looking forward to this analysis, aren't you?’, it gives the message prefix ‘Question’, the same prefix appears in repeated Discord message. Brawaru then shouts ‘Agent Coomer, report to Topside Tactical Operations Center.’, it has ‘Shout’ prefix, the same prefix appears in Discord. In the end Brawaru says ‘now I set message-types.local to “none”, local chat is no longer visible in discord’; which is true and the message doesn't appear in Discord.](https://user-images.githubusercontent.com/10401817/145730615-94195eea-e882-4b9f-897d-049a1fbee6e4.png)

Fixed #4989

### CODE REVIEW

File: EssentialsDiscord/src/main/java/net/essentialsx/discord/JDADiscordService.java [hunk 1]
```diff
@@ -303,6 +328,30 @@ public String parseMessageEmotes(String message) {
...
+                if (chatListener != null) {
+                    HandlerList.unregisterAll(chatListener);
+                }
+
+                chatListener = new BukkitChatListener(this);
+
+                Bukkit.getPluginManager().registerEvents(chatListener, plugin);
+            }
+        }
+    }
```
Comment (JRoy) [inline EssentialsDiscord/src/main/java/net/essentialsx/discord/JDADiscordService.java]:
```suggestion
    public void updateListener() {
        if (chatListener != null) {
            HandlerList.unregisterAll(chatListener);
            chatListener = null;
        }
        chatListener = (getSettings().isUseEssentialsEvents() && plugin.isEssentialsChat()) ? new EssentialsChatListener(this) : new BukkitChatListener(this);
        Bukkit.getPluginManager().registerEvents(chatListener, plugin);
    }
```

File: EssentialsDiscord/src/main/resources/config.yml [hunk 2]
```diff
@@ -46,6 +46,10 @@ channels:
 # players the essentials.discord.receive.primary permission.
 always-receive-primary: false
 
+# Whether to use Essentials Chat events instead of normal chat event.
+# This allows you to filter chat by its type (local, question, shout).
+use-essentials-events: false
```
Comment (JRoy) [inline EssentialsDiscord/src/main/resources/config.yml:51]:
```suggestion
use-essentials-events: true
```

File: EssentialsDiscord/src/main/java/net/essentialsx/discord/DiscordSettings.java [hunk 3]
```diff
@@ -107,6 +108,10 @@ public boolean isAlwaysReceivePrimary() {
         return config.getBoolean("always-receive-primary", false);
     }
 
+    public boolean isUseEssentialsEvents() {
+        return config.getBoolean("use-essentials-events", false);
```
Comment (JRoy) [inline EssentialsDiscord/src/main/java/net/essentialsx/discord/DiscordSettings.java:112]: Default to true here, users won't expect to have local chat relayed by default.

```suggestion
        return config.getBoolean("use-essentials-events", true);
```

File: EssentialsDiscord/src/main/resources/config.yml [hunk 4]
```diff
@@ -136,6 +140,15 @@ message-types:
   kick: staff
   # Message sent when a player's mute state is changed on the Minecraft server.
   mute: staff
+  # Message sent when a player talks in local chat.
+  # use-essentials-events must be set to "true" for this to work.
+  local: primary
```
Comment (JRoy) [inline EssentialsDiscord/src/main/resources/config.yml]: Users won't expect local chat to be relayed by default
```suggestion
  local: none
```

**Review by JRoy — CHANGES_REQUESTED** [review]
few things

**Comment by JRoy** [comment]
@brawaru could you update this to the latest 2.x so we can get this merged :)

**Comment by brawaru** [comment]
Rebased, but haven't tested anything yet. Will try to when I have time.

File: EssentialsDiscord/src/main/resources/config.yml [hunk 5]
```diff
@@ -46,6 +46,10 @@ channels:
 # players the essentials.discord.receive.primary permission.
 always-receive-primary: false
 
+# Whether to use Essentials Chat events instead of normal chat event.
+# This allows you to filter chat by its type (local, question, shout).
+use-essentials-events: true
```
Comment (JRoy) [inline EssentialsDiscord/src/main/resources/config.yml]:
```suggestion
use-essentials-events: false
```

File: EssentialsDiscord/src/main/java/net/essentialsx/discord/DiscordSettings.java [hunk 6]
```diff
@@ -107,6 +108,10 @@ public boolean isAlwaysReceivePrimary() {
         return config.getBoolean("always-receive-primary", false);
     }
 
+    public boolean isUseEssentialsEvents() {
+        return config.getBoolean("use-essentials-events", true);
```
Comment (JRoy) [inline EssentialsDiscord/src/main/java/net/essentialsx/discord/DiscordSettings.java]:
```suggestion
        return config.getBoolean("use-essentials-events", false);
```

**Review by JRoy — CHANGES_REQUESTED** [review]
i don't know why i requested these be enabled by default, they should be disabled 😄

**Review by JRoy — APPROVED** [review]

**Pull request merged**